Feature/dynamic weighted majority

AndreFCruz opened the pull request:
Changes proposed in this pull request:
- Implementation of the Dynamic Weighted Majority ensemble algorithm [1]

[1] Kolter and Maloof. Dynamic weighted majority: An ensemble method for drifting concepts. The Journal of Machine Learning Research, 8:2755-2790, December 2007. ISSN 1532-4435.

Loosely based on the MOA implementation of the same algorithm: https://github.com/Waikato/moa/blob/master/moa/src/main/java/moa/classifiers/meta/DynamicWeightedMajority.java

_Note:_ this is a different algorithm from #101 

@scikit-multiflow/scikit-multiflow-repo-admins

Checklist

- [x] Code complies with PEP-8 and is consistent with the framework.
- [x] Code is properly documented.
- [x] Tests are included for new functionality.
- [x] Travis CI build passes.
- [x] Test Coverage is maintained (threshold 0.2%)

jacobmontiel commented on src/skmultiflow/meta/dynamic_weighted_majority.py:
```diff
@@ -0,0 +1,271 @@
+            self.weight = weight
+
+        def __lt__(self, other):
+            self.weight < other.weight
```
Missing `return`?

```suggestion
            return self.weight < other.weight
```

jacobmontiel commented on src/skmultiflow/meta/dynamic_weighted_majority.py:
```diff
@@ -0,0 +1,271 @@
+            for compliance with the general skmultiflow interface.
+        """
+        self.epochs += 1
+        self.num_classes = max(
```
`self.num_classes` is not defined in the `__init__` method

AndreFCruz replied:
it's defined in the ```reset()``` method, which is called in ```__init__```

jacobmontiel commented on src/skmultiflow/meta/dynamic_weighted_majority.py:
```diff
@@ -0,0 +1,271 @@
+        for exp in self.experts:
+            exp.estimator.partial_fit(X, y, classes, weight)
+
+    def get_expert_predictions(self, X, classes=None, weight=None):
```
Parameters `classes` and `weight` are not used in this method

jacobmontiel commented on src/skmultiflow/meta/dynamic_weighted_majority.py:
```diff
@@ -0,0 +1,271 @@
+        weakest_expert_index = None
+        
+        for i, exp in enumerate(self.experts):
+            yHat = exp.estimator.predict(X)
```
The variable name should be lowercase `y_hat`

jacobmontiel commented on src/skmultiflow/meta/dynamic_weighted_majority.py:
```diff
@@ -0,0 +1,271 @@
+            self._construct_new_expert()
+        ]
+
+    def _scale_weights(self, maxWeight):
```
The variables names should be lowercase: `max_weight`, `scale_factor`

jacobmontiel commented on src/skmultiflow/meta/dynamic_weighted_majority.py:
```diff
@@ -0,0 +1,271 @@
+        """
+        Reset this ensemble learner.
+        """
+        self.epochs = 0
```
Some attributes are defined outside the `__init__` method:
`self.epochs`, `self.num_classes`, `self.experts`

AndreFCruz replied:
should I repeat the body of ```reset()``` in ```__init__``` ? (instead of calling reset in __init__)

jacobmontiel replied:
In my opinion, both are correct. Personally, I prefer calling `reset` from `__init__`.

jacobmontiel commented on tests/meta/test_dynamic_weighted_majority.py:
```diff
@@ -0,0 +1,45 @@
+                            0, 0, 1, 1, 1, 0, 0, 1, 1, 0, 1, 0, 0, 1, 1]
+    expected_correct_predictions = 44
+    expected_performance = 0.8979591836734694
+
```
Given that the method implements `predict_proba` please consider adding an explicit assertion for expected predicted probabilities.

Although `predict_proba` is called by `predict`, there could be cases where changes in `predict_proba` could be missed without an explicit test.

See for reference: https://github.com/scikit-multiflow/scikit-multiflow/blob/c4dbbb70d4ed839d95a18ca799f073ac9ff9ba49/tests/bayes/test_naive_bayes.py#L44-L46

AndreFCruz replied:
The ```predict_proba``` was a mistake, it was just to be used internally since the paper doesn't actually define what the class probability distributions would be.

I've addressed this in the most recent commit.

AndreFCruz commented:
Thanks for the feedback

A couple notes:
* some variables are not defined in ```__init__```, but are defined in the ```reset``` method which is called in ```__init__```
* ```predict_proba``` is no longer implemented
* removed the ```__lt__``` method from the inner class, this wasn't being used
* removed unused optional args
* changed variable naming to snake_case

jacobmontiel commented:
Thanks for the changes @AndreFCruz 

Regarding the instance attributes. This is just for readability, so people unfamiliar with the code can "see" all attributes in the `__init__`. For example, in this case:

``` python
    def __init__(self, n_estimators=5, base_estimator=NaiveBayes(),
                 period=50, beta=0.5, theta=0.01):
        """
        Creates a new instance of DynamicWeightedMajority.
        """
        super().__init__()

        self.max_experts = n_estimators
        self.base_estimator = base_estimator

        self.beta = beta
        self.theta = theta
        self.period = period

        # Following attributes are set later
        self.epoch = None
        self.num_classes = None
        self.experts = None

        self.reset()
```

Let me know what you think.

AndreFCruz commented:
Sure, I agree that improves readability.
Just committed the changes.
